### ISSUE TITLE
PKCS1 padding error on some EncryptionResponse packets

### ISSUE THREAD

**Issue by dannytech** [dannytech/Elytra#13]
Creating an issue to track this error. Unsure about the exact conditions which trigger this error. So far I know that when reading the EncryptionResponse packet the length fields for both the shared secret and verify token are 128 bytes. However, while the shared secret is read as the full 128 bytes, the verify token only ends up being 121 bytes, and it's in that case that the decryption fails.

```
(node:18293) UnhandledPromiseRejectionWarning: Error: error:0407109F:rsa routines:RSA_padding_check_PKCS1_type_2:pkcs decoding error
    at Object.privateDecrypt (internal/crypto/cipher.js:63:12)
    at Keypair.Decrypt (/workspaces/Elytra/src/protocol/Encryption.ts:22:16)
    at EncryptionResponsePacket.<anonymous> (/workspaces/Elytra/src/protocol/states/login/EncryptionResponsePacket.ts:62:54)
    at step (/workspaces/Elytra/src/protocol/states/login/EncryptionResponsePacket.ts:46:23)
    at Object.next (/workspaces/Elytra/src/protocol/states/login/EncryptionResponsePacket.ts:27:53)
    at /workspaces/Elytra/src/protocol/states/login/EncryptionResponsePacket.ts:21:71
    at new Promise (<anonymous>)
    at __awaiter (/workspaces/Elytra/src/protocol/states/login/EncryptionResponsePacket.ts:17:12)
    at EncryptionResponsePacket.Parse (/workspaces/Elytra/src/protocol/states/login/EncryptionResponsePacket.ts:72:16)
    at PacketFactory.<anonymous> (/workspaces/Elytra/src/protocol/PacketFactory.ts:200:26)
    at step (/workspaces/Elytra/src/protocol/PacketFactory.ts:33:23)
    at Object.next (/workspaces/Elytra/src/protocol/PacketFactory.ts:14:53)
    at /workspaces/Elytra/src/protocol/PacketFactory.ts:8:71
    at new Promise (<anonymous>)
    at __awaiter (/workspaces/Elytra/src/protocol/PacketFactory.ts:4:12)
    at PacketFactory.Parse (/workspaces/Elytra/src/protocol/PacketFactory.ts:158:16)
(Use `node --trace-warnings ...` to show where the warning was created)
(node:18293) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 2)
(node:18293) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
```

**Comment by dannytech** [comment]
The underlying buffer for ReadableBuffer shows that I'm not processing the entirety of the packet. Took a packet capture, which shows that the entire packet is being sent just fine over the wire, but when I create the ReadableBuffer, it truncates the packet. Surprised this hasn't caused more issues, have yet to find the exact location where this truncation happens.

```
Wireshark (working): 850201 8001 273369e6096f2839aa871a61f428781504026bd527956b327d7be0908c370074eb4592c31282b7ef3fcd19c070ba2494b0bf26272b9929ab30aeebb9109a89d12aa91f23f5747508dd5aae32184899eddc407228bc205648abeb6c35fe5f43823db599aec4e6192f28c65c8c93e43b950b056f3b54a9dedf43ed65c5c4ed91ca 8001 30bc1c76c452d119e4c9e5bffd5f8ea3ad2466df03368caf9b453e65d60c1215aa34ad2db96e05fdfa5761cd738a45181113b2ddaee316d4d88860a3428b3774b7734548baa25dbdd4ead31a6a826e9a6f7ae1265d8e159faf8c07c27434fd939bfc3e2f8175e251ec81bf04e3c8bec472d9ebea1a105034ef79cabb10db1908
Elytra (working):        01 8001 273369e6096f2839aa871a61f428781504026bd527956b327d7be0908c370074eb4592c31282b7ef3fcd19c070ba2494b0bf26272b9929ab30aeebb9109a89d12aa91f23f5747508dd5aae32184899eddc407228bc205648abeb6c35fe5f43823db599aec4e6192f28c65c8c93e43b950b056f3b54a9dedf43ed65c5c4ed91ca 8001 30bc1c76c452d119e4c9e5bffd5f8ea3ad2466df03368caf9b453e65d60c1215aa34ad2db96e05fdfa5761cd738a45181113b2ddaee316d4d88860a3428b3774b7734548baa25dbdd4ead31a6a826e9a6f7ae1265d8e159faf8c07c27434fd939bfc3e2f8175e251ec81bf04e3c8bec472d9ebea1a105034ef79cabb10db1908
Wireshark (error):   850201 8001 9843b072e305224d1a1b8197e3074549aa7d696346e1ca8aa6de179ded253339ffc4d0b4899925d7ed149bf8975eef72774b498be6f30bd936d4dbaea9ad56b9aec18be924b2c4eecd8b254b51a564a35cf84aeb43ae14b99f24da7d267bad694b961f9b8a3ad8f8577fde7db8cee616e06b31454d3bca1590ddcd10d58544ee 8001 7bb1dc937138c9eefa698dfe7c96c1b0bf7fff98ee6628e73c09e1c819e303fb70cb8e5926dd3fa006a64594d842957c5478588309056d9352cb0e06278d79120d868be2e6889afec9af0ba17e372453e28e667b27879577408c14c25f71047fc1d199ed7779ecaa1dcb5dd893b8ffa06b69211bd72a083f696c7fd25fa77376
Elytra (error):          01 8001 9843b072e305224d1a1b8197e3074549aa7d696346e1ca8aa6de179ded253339ffc4d0b4899925d7ed149bf8975eef72774b498be6f30bd936d4dbaea9ad56b9aec18be924b2c4eecd8b254b51a564a35cf84aeb43ae14b99f24da7d267bad694b961f9b8a3ad8f8577fde7db8cee616e06b31454d3bca1590ddcd10d58544ee 8001 7bb1dc937138c9eefa698dfe7c96c1b0bf7fff98ee6628e73c09e1c819e303fb70cb8e5926dd3fa006a64594d842957c5478588309056d9352cb0e06278d79120d868be2e6889afec9af0ba17e372453e28e667b27879577408c14c25f71047fc1d199ed7779ecaa1dcb5dd893b8ffa06b69211bd72a083f69
```

**Comment by dannytech** [comment]
Issue is all the way up `ClientBus.ts`.

https://github.com/dannytech/Elytra/blob/ade173afd9945883879c6f68078f42c539a91ef1/src/protocol/ClientBus.ts#L61-L65

As far as I can tell, the issue is that `client.on("data", ...)` receives a TCP data chunk, which is not guaranteed to be the full packet (it might be sent in two separate chunks. Not sure what to do about this. The issue is really that the length of a packet is not necessarily predictable, so I can't just say `if packetId == X then keep waiting for data until length == Y`. It might be that I have to figure out a way to stream the packet so it can keep receiving while being processed, and wait for further data if needed (read functions would have to block).

This also maybe explains why it hasn't happened before, most packets are not that large and so are just sent and received in one chunk.

**Comment by dannytech** [comment]
Side note, what this looks like is that the first chunk is as shown before in the Wireshark vs. Elytra comparison:

```
850201 8001 15de153f9d836bc5f18bf12546f42fae5f3a4d7c9f409ef985f390d3a4bed280e79d27297ca95e9881609a484f6e4d7cfd2c7303554ce6a4e11dac5dd1b48c4cb9d601a882f2eaa44f596083d98f062dbec078023eb2ba9787b6a3fbaee5f8d38be0f6a966e3ed2744069ceb1e1410786572ba9c88a1c652c96012fb21cc3cc4 8001 8ea0f88d7a3b41d2245a9217b3b8065d600393d2b635bc1b759e031f9b1aa3b1b57f10e516a78730d9bdfcae6ac087679c4989df979aa569456124f415faa48eb2c5770df0f2ce2a504c50d5d2c5bc0607c910d069ae7bcdeb54f5fc94f78b84a761945de3143ad8e1459bafc92507a0d38063e8fb627b1620
```

But then, the server processes a new "packet", which is actually the remainder of the bytes. In total, the complete packet with these two parts is 263 bytes long, as expected (two 128-byte ciphertexts + a 1-byte packet ID + a 2-byte packet length + two 2-byte array lengths).

```
edf2393a0bc9a4
```

**Comment by dannytech** [comment]
Working on fix in https://github.com/dannytech/Elytra/tree/issues/13.

**Comment by dannytech** [comment]
![](https://github.trello.services/images/mini-trello-icon.png) [Fix #13 (TCP chunk length issues)](https://trello.com/c/aZDLt7BI/48-fix-13-tcp-chunk-length-issues)

**Comment by dannytech** [comment]
Quick update: rewrite of packet processing in process. The initial stage is now handled using a stream which blocks until the expected packet length is received. Preparing to commit, after which the next step is to fix how legacy server list ping works with that setup (simply removing it is not really an option, because then the client simply has junk data left over in the stream). Streams are a pain to learn but very cool, at this point I regret that they can't be fully utilized because the entire stream is not compressed (not to mention the Adler32 checksum has to be confirmed before the packet can be processed). 

Still todo:

- [] Fix legacy server list ping
- Set a timeout on stream read operations?

**Comment by dannytech** [comment]
Fixed in 97f03c7.

https://github.com/dannytech/Elytra/blob/issues/13/src/protocol/ProtocolStub.ts#L25-L35
